### PR TITLE
`General`: Redesign learning goals and store progress

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The learning goals feature was rarely used, because it only offered limited benefits to the students and instructor.

### Description

In this pull request, we renamed the _learning goals_ to _competencies_ in the client.
The user interface has been overhauled and now features three circular progress bars to track the progress (% of completed lecture units), confidence (% average score w.r.t the mastery threshold), and mastery (weighted metric of progress and confidence).
A new competency details page allows students to work on the specific course material towards achieving a specific competency.

Exercises and lecture units can now be linked to competencies on the respective management pages (instead of only on the competency edit screen).
The progress for all students in each competency is now persisted to the database in preparation for future work on learning paths.
Updating the progress always happens asynchronously and does not delay the server's response. It is triggered when a new result for an exercise is processed in the participant score scheduler, when the user completes a lecture unit, or when a learning goal/lecture unit is edited by an instructor. The user can also manually update the progress by clicking on the “refresh” button.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Course with some Exercises, a Lecture and Lecture Units

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a Competency
4. Go to any lecture unit or exercise and add it to the competency
5. Submit something to the exercise and grade it, the competency progress should update
6. Complete some lecture units, the competency progress should update

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with all exercise types

1. Log in to Artemis
2. Participate in the exam as a student
3. Make sure that the functionality of all exercises is unchanged

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

### [Test Coverage](https://app.codecov.io/gh/ls1intum/Artemis/pull/6050)

| Class/File | Branch | Line |
|------------|-------:|-----:|
| fireworks.component.ts | 100% | 100% |
| learning-goal-rings.component.ts | 100% | 100% |
| learning-goal-selection.component.ts | 84% | 87% |
| learning-goal-card.component.ts | 76% | 93% |
| course-learning-goals-details.component.ts | 76% | 83% |
| course-learning-goals.component.ts | 76% | 97% |
| learningGoal.service.ts | 72% | 92% |
| LearningGoalProgressService.java | 68% | 89% |
| LearningGoalService.java | 100% | 100% |
| LearningGoalResource.java | 80% | 97% |
| CourseService.java | 80% | 92% |

### Screenshots

course-lgs

lg-details

lgs-mgmt

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java [hunk 1]
```diff
@@ -74,6 +79,14 @@ public void setDescription(String description) {
         this.description = description;
     }
 
+    public Integer getMasteryThreshold() {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java:82]: This can be an int since null can no longer get returned
```suggestion
    public int getMasteryThreshold() {
```

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+     * @param course The course for which to fetch the learning goals from
+     * @return All learning goals of the course with the updated progress for the user
+     */
+    public Set<LearningGoal> getLearningGoalsAndUpdateProgressByUserInCourse(User user, Course course) {
+        return learningGoalRepository.findAllForCourse(course.getId()).stream().peek(learningGoal -> {
+            var updatedProgress = updateLearningGoalProgress(learningGoal.getId(), user);
+            if (updatedProgress != null) {
+                learningGoal.setUserProgress(Set.of(updatedProgress));
+            }
+        }).collect(Collectors.toSet());
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java:115]: Please avoid using peek. From the [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer-): "API Note: This method exists mainly to support debugging, where you want to see the elements as they flow past a certain point in a pipeline."
```suggestion
        var learningGoals = learningGoalRepository.findAllForCourse(course.getId());
        learningGloals.forEach(learningGoal -> {
            var updatedProgress = updateLearningGoalProgress(learningGoal.getId(), user);
            if (updatedProgress != null) {
                learningGoal.setUserProgress(Set.of(updatedProgress));
            }
        });
        return learningGoals;
```

File: src/main/java/de/tum/in/www1/artemis/domain/lecture/LectureUnit.java [hunk 3]
```diff
@@ -55,13 +53,15 @@ public abstract class LectureUnit extends DomainObject implements Completable {
...
+    @ManyToMany
+    @JoinTable(name = "learning_goal_lecture_unit", joinColumns = @JoinColumn(name = "lecture_unit_id", referencedColumnName = "id"), inverseJoinColumns = @JoinColumn(name = "learning_goal_id", referencedColumnName = "id"))
     @OrderBy("title")
+    @JsonIgnoreProperties({ "lectureUnits", "course" })
     @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
     protected Set<LearningGoal> learningGoals = new HashSet<>();
 
     @OneToMany(mappedBy = "lectureUnit", fetch = FetchType.LAZY, cascade = CascadeType.REMOVE, orphanRemoval = true)
-    @JsonIgnore
+    @JsonIgnore // important!
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/domain/lecture/LectureUnit.java:64]: Maybe include why it's important

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 4]
```diff
@@ -753,7 +753,7 @@ private List<StudentParticipation> filterParticipationsWithRelevantResults(List<
                 .filter(participation -> participation.getParticipant() != null)
 
                 // filter all irrelevant results, i.e. rated = false or no completion date or no score
-                .peek(participation -> {
+                .map(participation -> {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:756]: Why this change?
Maybe we could change this whole block to directly calling removeIf on the list and then simply iterating over the entries that are left in the list

File: src/main/java/de/tum/in/www1/artemis/domain/ILearningObject.java [hunk 5]
```diff
@@ -2,8 +2,9 @@
 
 import java.time.ZonedDateTime;
 import java.util.Optional;
+import java.util.Set;
 
-public interface Completable {
+public interface ILearningObject {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/domain/ILearningObject.java:7]: Seams like you only allow certain inheritance from this interface further down. You might want to think about marking this interface as sealed, which is a new feature of Java 17

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            Set<LearningGoal> learningGoals;
+            if (learningObject instanceof Exercise exercise) {
+                learningGoals = exerciseRepository.findByIdWithLearningGoals(exercise.getId()).map(Exercise::getLearningGoals).orElse(null);
+            }
+            else if (learningObject instanceof LectureUnit lectureUnit) {
+                learningGoals = lectureUnitRepository.findByIdWithLearningGoals(lectureUnit.getId()).map(LectureUnit::getLearningGoals).orElse(null);
+            }
+            else {
+                throw new IllegalArgumentException("Learning object must be either LectureUnit or Exercise");
+            }
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java:136]: Since this is now repeated, you could move this as default method to the ILearningObject

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+     * Updates the progress value (and confidence score) of the given learning goal and user, then returns it
+     * @param learningGoalId The id of the learning goal to update the progress for
+     * @param user The user for which the progress should be updated
+     * @return The updated learning goal progress, which is also persisted to the database
+     */
+    public LearningGoalProgress updateLearningGoalProgress(Long learningGoalId, User user) {
+        var learningGoal = learningGoalRepository.findByIdWithExercisesAndLectureUnitsAndCompletions(learningGoalId).orElse(null);
+
+        if (user == null || learningGoal == null) {
+            // If the user or learning goal no longer exist, there is nothing to do
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java:165]: This is more or less the same as the debug statement right below

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            }
+            else if (learningObject instanceof LectureUnit lectureUnit) {
+                learningGoals = lectureUnitRepository.findByIdWithLearningGoals(lectureUnit.getId()).map(LectureUnit::getLearningGoals).orElse(null);
+            }
+            else {
+                throw new IllegalArgumentException("Learning object must be either LectureUnit or Exercise");
+            }
+
+            if (learningGoals == null) {
+                // Learning goals couldn't be loaded, the exercise/lecture unit might have already been deleted
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java:139]: This is more or less the same as the debug statement right below

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+            logger.debug("User or learning goal no longer exist, skipping.");
+            return null;
+        }
+
+        var learningGoalProgress = learningGoalProgressRepository.findEagerByLearningGoalIdAndUserId(learningGoalId, user.getId());
+
+        if (learningGoalProgress.isPresent()) {
+            var lastModified = learningGoalProgress.get().getLastModifiedDate();
+            if (lastModified != null && lastModified.isAfter(Instant.now().minusSeconds(1))) {
+                // If we have updated the progress within the last seconds, skip it
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java:175]: This is more or less the same as the debug statement right below

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+                logger.debug("Learning goal progress has been updated very recently, skipping.");
+                return learningGoalProgress.get();
+            }
+        }
+
+        // Now do the heavy lifting
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java:181]: _Interesting_ (but not very helpful) comment

File: src/main/webapp/app/overview/course-learning-goals/course-learning-goals-details.component.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+    get confidence(): number {
+        // Confidence level (average score in exercises) in proportion to the threshold value (max. 100 %)
+        // Example: If the student’s latest confidence level equals 60 % and the mastery threshold is set to 80 %, the ring would be 75 % full.
+        return Math.min(Math.round(((this.getUserProgress().confidence ?? 0) / (this.learningGoal.masteryThreshold ?? 100)) * 100), 100);
+    }
+
+    get mastery(): number {
+        // Advancement towards mastery as a weighted function of progress and confidence
+        const weight = 2 / 3;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/course-learning-goals/course-learning-goals-details.component.ts:98]: This weight is used in multiple places, should we convert this to a global constant?

File: src/test/javascript/spec/component/course-registration/course-prerequisites-modal.component.spec.ts [hunk 12]
```diff
@@ -3,22 +3,14 @@ import { ArtemisTranslatePipe } from 'app/shared/pipes/artemis-translate.pipe';
 import { MockPipe, MockProvider } from 'ng-mocks';
 import { LearningGoalService } from 'app/course/learning-goals/learningGoal.service';
 import { of } from 'rxjs';
-import { LearningGoal } from 'app/entities/learningGoal.model';
+import { LearningGoal, LearningGoalProgress } from 'app/entities/learningGoal.model';
```
Comment (JohannesStoehr) [inline src/test/javascript/spec/component/course-registration/course-prerequisites-modal.component.spec.ts:6]: Unused imports

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Looked through the code and left some comments

File: src/main/java/de/tum/in/www1/artemis/domain/ILearningObject.java (hunk 5 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/domain/ILearningObject.java:7]: Thanks for pointing this out! I'd have to move the classes to the same package for this (_"Class is not allowed to extend sealed class from another package"_), but I like the current structure.

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java (hunk 6 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java:136]: I'd rather not have so much service code in domain objects.

File: src/main/webapp/app/overview/course-learning-goals/course-learning-goals-details.component.ts (hunk 11 above)
Comment (ge65cer) [reply src/main/webapp/app/overview/course-learning-goals/course-learning-goals-details.component.ts:98]: I am thinking about moving this calculation to the server in a follow-up (because we recently discussed a similar issue with score calculation in the client w.r.t. the mobile app development).

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts [hunk 13]
```diff
@@ -83,6 +85,7 @@ export class AttachmentUnitFormComponent implements OnInit, OnChanges {
             releaseDate: [undefined as dayjs.Dayjs | undefined],
             version: [1],
             updateNotificationText: [undefined as string | undefined, [Validators.maxLength(1000)]],
+            learningGoals: [undefined as LearningGoal[] | undefined],
```
Comment (simon-hng) [inline src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts:88]: How about creating a separate type definition of this group instead of casting `undefined` into the wanted types

File: src/main/webapp/app/overview/course-learning-goals/course-learning-goals-details.component.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        return Math.round((1 - weight) * this.progress + weight * this.confidence);
+    }
+
+    get isMastered(): boolean {
+        return this.mastery >= 100;
+    }
+
+    completeLectureUnit(event: LectureUnitCompletionEvent): void {
+        if (event.lectureUnit.lecture && event.lectureUnit.visibleToStudents && event.lectureUnit.completed !== event.completed) {
```
Comment (simon-hng) [inline src/main/webapp/app/overview/course-learning-goals/course-learning-goals-details.component.ts:107]: negate that if condition and return to decrease level of nesting

File: src/main/webapp/app/overview/course-learning-goals/course-learning-goals.component.html [hunk 15]
```diff
@@ -1,41 +1,44 @@
...
+                </div>
+                <div *ngIf="!isCollapsed">
+                    <jhi-learning-goal-card
+                        *ngFor="let prerequisite of prerequisites; index as i; trackBy: identify"
+                        [learningGoal]="prerequisite"
+                        [courseId]="courseId"
+                        [isPrerequisite]="true"
+                    ></jhi-learning-goal-card>
+                </div>
+                <div class="collapsed"></div>
```
Comment (simon-hng) [inline src/main/webapp/app/overview/course-learning-goals/course-learning-goals.component.html:26]: What's the reason for this empty div

**Review by JohannesStoehr — APPROVED** [review]
Thanks for addressing my comments. Code looks good overall

**Review by simon-hng — COMMENTED** [review]
Code looks good. Left some minor remarks

File: src/main/webapp/app/entities/learningGoal.model.ts [hunk 16]
```diff
@@ -2,27 +2,48 @@ import { BaseEntity } from 'app/shared/model/base-entity';
 import { Course } from 'app/entities/course.model';
 import { Exercise } from 'app/entities/exercise.model';
 import { LectureUnit } from 'app/entities/lecture-unit/lectureUnit.model';
+import { IconProp } from '@fortawesome/fontawesome-svg-core';
+import { faBrain, faComments, faCubesStacked, faMagnifyingGlass, faPenFancy, faPlusMinus, faQuestion } from '@fortawesome/free-solid-svg-icons';
 
 /**
  * The available learning goal types (based on Bloom's Taxonomy)
  */
 export enum LearningGoalTaxonomy {
```
[the remark this answers is not in the document]
Comment (simon-hng) [reply src/main/webapp/app/entities/learningGoal.model.ts:11]: But yeah, mostly minor and possibly a follow-up task

File: src/main/webapp/app/overview/course-learning-goals/course-learning-goals.component.html (hunk 15 above)
Comment (ge65cer) [reply src/main/webapp/app/overview/course-learning-goals/course-learning-goals.component.html:26]: Adds a border to the bottom (see also course lecture page).

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts (hunk 13 above)
Comment (ge65cer) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts:88]: I don't think this makes sense for this very specific use case in the Angular Forms definition.

File: src/main/webapp/app/entities/learningGoal.model.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (ge65cer) [reply src/main/webapp/app/entities/learningGoal.model.ts:11]: If we currently use Enums everywhere, if we decide against that we should maybe refactor all in one go in a separate PR.

**Review by simon-hng — DISMISSED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Re-approve

**Review by Hialus — CHANGES_REQUESTED** [review]
Tested on TS10.

When creating a text lecture unit I have this empty field that I can interact with in some way, but doesn't do anything:
grafik
This is before and after I created a competency.
Only after I added a competency to the unit I get this hard to read result:
grafik
That I can for some reason interact with even though it changes nothing?
grafik
Please fix this box to be more intuitive to understand and make it non interactable if it does not actually do anything.


I also ran into a weird server error when completing a text lecture unit and put the corresponding log here: https://gist.github.com/Hialus/4224b8e625c730cc04bc2b0f2fb7dd64
I am not able to reproduce this, so it was likely some weird edge case (maybe on first completion?), but maybe you can add a check against this either way.

The competency score updates accordingly, but since I am not really familiar with this system I cannot say if it updates correctly. 
One thing I noticed here: When I submit to a programming exercise while observing the competency detail page this happens:
- The programming exercise status changes to "Building and testing" automatically
- The programming exercise status changes to the actual feedback automatically
- The competency graph does not update automatically and I have to manually refresh the page
This is a somewhat niche use case, but it would be great if you could still fix this.

The database migration seems to work properly on a Postgres Testserver though!

**Comment by ge65cer** [comment]
@Hialus This is a standard [“select” input field](https://getbootstrap.com/docs/5.3/forms/select/). It seems the color theming for dark mode is incorrect, which will be addressed.

**Comment by Hialus** [comment]
> @Hialus This is a standard [“select” input field](https://getbootstrap.com/docs/5.3/forms/select/). It seems the color theming for dark mode is incorrect, which will be addressed.

Alright. But what am I selecting the competencies for? I don't see anything else that I can do with them after selecting them. That was what I am referring to.

**Review by simon-hng — APPROVED** [review]

**Review by simon-hng — APPROVED** [review]

**Review by Hialus — DISMISSED** [review]
Dark mode issue is fixed now. I still don't understand why the competency list is selectable, but its not an important issue.
As I said automatic updating is niche and is more of a nice-to-have feature.

So from what I can tell this PR works as expected. 👍🏻

**Comment by krusche** [comment]
@ge65cer could you add the test coverage to the PR description please?

**Comment by ge65cer** [comment]
@krusche Yes, I will as soon as the server tests work on develop and I can update this branch.

**Comment by krusche** [comment]
@ge65cer tests should be more stable again on `develop` 
(in case you detect another flaky test, please notify me on Slack)

**Comment by ge65cer** [comment]
@krusche Tests run through, coverage table was appended.

File: src/main/java/de/tum/in/www1/artemis/repository/LearningGoalProgressRepository.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    Optional<LearningGoalProgress> findByLearningGoalIdAndUserId(@Param("learningGoalId") Long learningGoalId, @Param("userId") Long userId);
+
+    @Query("""
+            SELECT learningGoalProgress
+            FROM LearningGoalProgress learningGoalProgress
+            LEFT JOIN FETCH learningGoalProgress.user
+            LEFT JOIN FETCH learningGoalProgress.learningGoal
+            WHERE learningGoalProgress.learningGoal.id = :learningGoalId
+            AND learningGoalProgress.user.id = :userId
+            """)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/LearningGoalProgressRepository.java:48]:
```suggestion
    @Query("""
            SELECT p
            FROM LearningGoalProgress p
                LEFT JOIN FETCH p.user
                LEFT JOIN FETCH p.learningGoal
            WHERE p.learningGoal.id = :learningGoalId
                AND p.user.id = :userId
            """)
```

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/repository/LearningGoalProgressRepository.java:48]: please indent the left joins and the logical operations below the "where" statement, so that the query is easier to read. Also use shorter table variables (instead of p, you can also use lgp e.g. in this case)

**Review by krusche — CHANGES_REQUESTED** [review]
I looked over the server changes, and they mostly look good from what I can say in a code review on Github. Could I kindly ask you to improve the indentation of all queries? I added one example of the preferred style which we would like to enforce in the future.

In addition, I would kindly ask you to comment on the performance of the approach in the PR description. It is currently unclear to me which actions trigger a re-calculation of the learning goal progress. I would assume a new exercise result would be relatively simple and fast to store so that students would not notice a slower performance. However, what happens when instructors change competencies when there are already existing scores? What happens if instructors re-evaluate an exercise? Will all those actions automatically update the learning goal progress?

I really like the new user interface and would kindly ask you to add screenshots (potentially the same as used in the PR description) to the documentation on https://docs.artemis.cit.tum.de/user/learning-analytics/ and to explain them to end user in a follow-up PR.

**Comment by krusche** [comment]
One additional comment. It would be awesome, if we can merge the PR as part of the 6.0 release (currently planend on Friday, January 20). I don't see any obstacles in my code review, however it would be nice, if some testers could re-approve the PR.

**Comment by ge65cer** [comment]
> In addition, I would kindly ask you to comment on the performance of the approach in the PR description. It is currently unclear to me which actions trigger a re-calculation of the learning goal progress. I would assume a new exercise result would be relatively simple and fast to store so that students would not notice a slower performance. However, what happens when instructors change competencies when there are already existing scores? What happens if instructors re-evaluate an exercise? Will all those actions automatically update the learning goal progress?

A new exercise result triggers a progress update, which is taken care of right in the participant score scheduler (which runs in a separate thread). This also applies to re-evaluating an exercise and happens during the participant score update as well.
When the instructor edits a competency and changes the linked learning objects, this will trigger an asynchronous update only on existing progress entities of this learning goal.
The user can always manually update the competency progress as well using the “refresh” button.

**Review by JohannesStoehr — APPROVED** [review]
Checked latest commits for code issues, re-approve

**Review by danielbode — DISMISSED** [review]
Tested on ts10. Works as expected and I really like the new design! The code also looks fine.

**Review by stefanwaldhauser — DISMISSED** [review]
Great work! Works as expected and code looks good too!

**Review by Hialus — APPROVED** [review]
Reapprove (limited) manual testing.

**Review by maximiliansoelch — APPROVED** [review]
Tested on TS10, works as expected! 
Great work! 👍

**Comment by ge65cer** [comment]
✅ Cypress, client, and server tests passed. Five approvals from developers.

**Review by TheZoker — APPROVED** [review]
Tested on TS10 and LGTM 👍

**Review by LiaLism — APPROVED** [review]
Everything works as expected.
